capstone/src/main/java/com/google/sps/data/CachedInterestVector.java

```diff
@@ -0,0 +1,79 @@
+package com.google.sps.data;
+ 
+import com.google.sps.data.Room;
+import com.google.sps.data.User;
+import com.google.sps.data.Tag;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.lang.Iterable;
+import java.lang.Math;
+ 
+public class CachedInterestVector {
+
+    public static HashMap<String, Double> embeddedEntityToHashMap(EmbeddedEntity embeddedVectorMap) {
+        HashMap<String, Double> vectorHashMap = new HashMap<>();
+        for(String tag : embeddedVectorMap.getProperties().keySet()) {
+            vectorHashMap.put(tag, (Double) embeddedVectorMap.getProperty(tag));
+        }
+        return vectorHashMap;
+    }
+
+    public static EmbeddedEntity hashMapToEmbeddedEntity(HashMap<String, Double> vectorHashMap){
+        EmbeddedEntity embeddedVectorMap = new EmbeddedEntity();
+        for(String tag : vectorHashMap.keySet()) {
+            embeddedVectorMap.setProperty(tag, vectorHashMap.get(tag));
+        }
+        return embeddedVectorMap;
+    }
+
+    public static void denormalizeVectorHashMap(HashMap<String, Double> vectorHashMap, Double magnitude) {
```

Comment:
This would be better to return a completely separate HashMap instead of modifying the existing one in place. A couple benefits of this are 1. You can improve typing by returning a `HashMap<String, Integer>` (which will also change the signature of some of the other functions) since that is more representative of what a denormed map should be. 2. If somehow a bug cropped up and this threw an exception, then the map would be in a mutant half-normalized state. This is probably pretty unlikely in this particular function, but the practice of modifying things in place and using that as the "return" value even though its just a parameter is not best practice because it can be hard to spot this problem.